### PR TITLE
[gloo] dont hold locks in calls to buffer in ProcessGroupGloo:RecvWork::wait() and
SendWork::wait()

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33926 [gloo] dont hold locks in calls to buffer in ProcessGroupGloo:RecvWork::wait() and
SendWork::wait()**
SendWork::wait()**

Closes https://github.com/pytorch/pytorch/issues/30164
The UnboundBuffer calls here are already protected by a mutex. We only
need to hold the lock while writing the shared structures completed_ and
exception_.

Differential Revision: [D20154546](https://our.internmc.facebook.com/intern/diff/D20154546/)

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 1]
```diff
@@ -333,13 +333,15 @@ ProcessGroupGloo::SendWork::SendWork(
 
 bool ProcessGroupGloo::SendWork::wait() {
   bool sendCompleted = false;
-  std::unique_lock<std::mutex> lock(mutex_);
   try {
     sendCompleted = buffer_->waitSend();
   } catch (...) {
+    // Lock to write exception_
+    std::unique_lock<std::mutex> lock(mutex_);
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: Let's make them lock guard to be explicit that the following code will not manually unlock them?

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 2]
```diff
@@ -333,13 +333,15 @@ ProcessGroupGloo::SendWork::SendWork(
...
     exception_ = std::current_exception();
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: Can we avoid having two locks here by having a function local exception variable and only acquire the lock after the catch? Will it be better if we have a finally block and acquire lock there? sth like:

```c++
try {
  sendCompleted = buffer_->waitSend();
} catch (...) {
  exception = std::current_exception();
} finally {
  std::lock_guard<std::mutex> lock(mutex_);
  completed_ = true;
  if (exception) {
    std::swap(exception, exception_);
    std::rethrow_exception(exception_);
  }
}

return sendCompleted;
```

**Review by mrshenli — APPROVED** [review]
Thanks for fixing! Left some minor comments inline. Stamp to unblock.

File: torch/lib/c10d/ProcessGroupGloo.cpp (hunk 2 above)
Comment (rohan-varma) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: I think we can add the function local variable, although C++ does not support `finally` for the 2nd suggestion.

**Pull request closed without merging**